module/zfs/spa.c

```diff
@@ -2316,6 +2320,186 @@ vdev_count_verify_zaps(vdev_t *vd)
+
+	if (hostid == myhostid)
+		return (B_FALSE);
+
+	if (config_pool_state != POOL_STATE_ACTIVE)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+/*
+ * Perform the import activity check.  If user canceled import or we
+ * detected activity then fail.
+ */
+static int
+spa_activity_check(spa_t *spa, uberblock_t *ub)
+{
+	uint64_t import_intervals = MAX(zfs_mmp_import_intervals, 1);
+	uint64_t txg = ub->ub_txg;
+	uint64_t timestamp = ub->ub_timestamp;
+	uint64_t import_delay = 0;
+	hrtime_t import_expire;
+	nvlist_t *mmp_label = NULL;
+	vdev_t *rvd = spa->spa_root_vdev;
+	kcondvar_t cv;
+	kmutex_t mtx;
+	int error = 0;
+
+	cv_init(&cv, NULL, CV_DEFAULT, NULL);
+	mutex_init(&mtx, NULL, MUTEX_DEFAULT, NULL);
+	mutex_enter(&mtx);
+
+	/*
+	 * Preferentially use the zfs_mmp_interval from the node which last
+	 * imported the pool.  This value is stored in an MMP uberblock as.
+	 *
+	 * ub_mmp_delay * vdev_count_leaves() == zfs_mmp_interval
+	 */
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay)
+		import_delay = import_intervals * ub->ub_mmp_delay *
+		    vdev_count_leaves(spa);
+
+	/* Apply a floor using the local default values. */
+	import_delay = MAX(import_delay, import_intervals *
+	    MSEC2NSEC(zfs_mmp_interval));
+
+	/* Add a small random factor in case of simultaneous imports (0-25%) */
+	import_expire = gethrtime() + import_delay +
+	    (import_delay * spa_get_random(250) / 1000);
+
+	while (gethrtime() < import_expire) {
+		vdev_uberblock_load(rvd, ub, &mmp_label);
+
+		if (txg != ub->ub_txg || timestamp != ub->ub_timestamp) {
+			error = SET_ERROR(EREMOTEIO);
+			break;
+		}
+
+		if (mmp_label) {
+			nvlist_free(mmp_label);
+			mmp_label = NULL;
+		}
+
+		error = cv_timedwait_sig_hires(&cv, &mtx, NANOSEC, 1, 0);
+		if (error != -1) {
+			error = SET_ERROR(EINTR);
+			break;
+		}
+		error = 0;
+	}
+
+	mutex_exit(&mtx);
+	cv_destroy(&cv);
+
+	/*
+	 * If the pool is active store the hostname and hostid found in the
+	 * active configuration in spa->spa_load_info.  This allows for
+	 * a descriptive error message to be generated for the adminstrator.
+	 *
+	 * ZPOOL_CONFIG_IMPORT_HOSTNAME - hostname from the active pool
+	 * ZPOOL_CONFIG_IMPORT_HOSTID - hostid from the active pool
+	 */
+	if (error == EREMOTEIO) {
+		char *hostname = "unknown";
+		uint64_t hostid = 0;
+
+		if (mmp_label) {
+			if (nvlist_exists(mmp_label, ZPOOL_CONFIG_HOSTNAME))
+				hostname = fnvlist_lookup_string(mmp_label,
+				    ZPOOL_CONFIG_HOSTNAME);
+
+			if (nvlist_exists(mmp_label, ZPOOL_CONFIG_HOSTID))
+				hostid = fnvlist_lookup_uint64(mmp_label,
+				    ZPOOL_CONFIG_HOSTID);
+
+			fnvlist_add_string(spa->spa_load_info,
+			    ZPOOL_CONFIG_IMPORT_HOSTNAME, hostname);
+			fnvlist_add_uint64(spa->spa_load_info,
+			    ZPOOL_CONFIG_IMPORT_HOSTID, hostid);
+		}
+
+		cmn_err(CE_WARN, "pool '%s' could not be loaded as it is being "
+		    "accessed by another system (host: %s hostid: 0x%lx)",
+		    spa_name(spa), hostname, (unsigned long)hostid);
```

Comment:
We should use the pool name provided in the passed `config` here to avoid logging the generic `$import` pool name which is set for tryconfig.